### ISSUE TITLE
new search algorithm failed with random condition

### ISSUE THREAD

**Issue by ckxz105** [usc-isi-i2/dsbox-ta2#215]
Currently our program with new search algorithm in master branch seems will end randomly on different conditions.
I tried to run the program 3 times we exactly same input, but get 3 different results:
1. First time only `Large_column_number_with_numerical_only_classif` generated results.
![image](https://user-images.githubusercontent.com/25087508/46446134-5f3ca380-c72f-11e8-9143-608e2f4008c7.png)

2. Second time 2 template generated results
![image](https://user-images.githubusercontent.com/25087508/46446120-4b913d00-c72f-11e8-81a8-9082f45574b8.png)
3. Third time only naive ayes generated results
![image](https://user-images.githubusercontent.com/25087508/46446128-58ae2c00-c72f-11e8-8719-4f158c5f3c67.png)

It seems because of the different initial random condition?

**Comment by RqS** [comment]
#207 Seems like same thing

**Comment by proska** [comment]
Please check the [https://github.com/usc-isi-i2/dsbox-ta2/blob/master/python/dsbox/controller/controller.py](here) at lines 401, 420, and 439. We are passing a "num_iter" to each one of the search methods which indicates how many rounds of search each do. 
For random searches, it indicates the total number of pipeline samples, and for dimensional search, it indicates how many templates will be sampled.
To sum it up, the code on master branch basically runs a parallel random search with only one sample pipeline that was used for easy testing and development. If you want to evaluate the whole system just increase the number. The random searches are pretty stable and dimensional ones will be fixed by the end of this week.

**Comment by proska** [comment]
I am still working on a solution for the bug, but I thought to share some progress here:
The problem seems to be when each worker tries to push the evaluation results back into the shared queue. That is where they stuck, hence the whole controller stops any progress. (A dirty workaround was to add a timeout and ignore the worker). 
To pinpoint the reason for our problem, my first guess was serialization that will happen inside the queue. So I added a pickling/unpickling step just before pushing the results into the queue in each worker, which seem to fix the issue in my test cases (still needs extensive testing on dsbox server). 
Before pushing, I noticed that I was working on the master branch and there has been a bunch of commits in the development branch which needed to be merged with my fix. However after merging the whole code seem to break on the evaluation. To be more specific, the code throws an exception on lines 114 and 118 (Result object is not iterable). 
Does anyone have any idea about it? I also noticed a FIXME comment on both lines which is not clear to me.

**Comment by proska** [comment]
I have pushed a fix for this issue. In case you still face it, please send me the log dump of your execution.

**Comment by proska** [comment]
Seems to be fixed after jobmanager bug fix
